### PR TITLE
Name operation segment earlier

### LINKED ISSUE #117: Generic POST//graphql appears after running web server for a while
## Description

Hey all, I'm running into a weird issue and was wondering if this is the right place to ask. A couple days ago, I integrated this plugin with our new relic integration (we use Apollo Server with express). Everything appeared to be working, but I noticed that after our servers run for a couple hours, performance seems to "degrade" according to new relic. When I look at the top most time consuming transaction, I see that it's a generic `post` to `/graphql`. 

perf-issues-screenshot

When I dive into the segment trace for this generic post, I see that it's the amalgamation of all our graphql resolvers (this continued for many more segments) 

Screen Shot 2021-07-14 at 5 24 05 PM

What I find even odder is that when we deploy and new code starts running on our servers, is that this generic post on our top most time consumings transactions completely goes away for a couple hours and then comes back. (Screenshot below from 30 mins after fresh deploy). 

non-perf-issues-screenshot

As far as I can tell, nothing in our code calls all the resolvers at the same time. So my question is, is this expected behavior? Are we doing something wrong on our side? If it is expected,  why wouldn't this new relic transaction have an end depth like all the others? Also if so, is there a way to ignore it if it's not actually affecting perf? My goal here is to figure out if this is actually a server perf issue and fix it or it ignore it depending on what I learn here. Many thanks in advance. 

- Alex

## Expected Behavior

Generic `post//graphql` to not show up in our new relic transactions and instead show each individual transaction. 

## Steps to Reproduce

Install new relic, use Apollo express, add this as a plugin (as far as I can tell). 

## Your Environment

newrelic version 6.13.2
newrelic/apollo-server-plugin version  0.3.0
Node Version v12.13.1
Running on [heroku stack 18](https://devcenter.heroku.com/articles/heroku-18-stack)

### PR DESCRIPTION
**Pull request by bizob2828**
## Proposed Release Notes
 * Moved naming of operation segment and updating transaction to `validationDidStart` phase of Apollo Server request.

## Links
Closes #117

## Details
I was able to reproduce the bug #117 by employing the brute force method @carlo-808 mentioned with `connect-timeout`.  So it appears if you initiate an apollo server request but throw some error out of band where apollo does not emit the `didEncounterErrors` hook you get into this situation.  So in the `connect-timeout` case, that middleware will emit a timeout after n milliseconds.  That timeout is not being caught by apollo thus not allowing `didEncounterErrors` and subsequently `willSendResponse` hooks to fire. We were updating the name of the operation segment and transaction in the `willSendResponse`.  There was no reason aside from we had the most information but the only information we need now that we refactored finding the deepest path and naming is from the graphql document AST.  The first time this is available is in `validationDidStart`, hence the PR.

### CODE REVIEW

File: tests/unit/create-plugin.test.js [hunk 1]
```diff
@@ -63,16 +63,12 @@ tap.test('createPlugin edge cases', (t) => {
...
     const responseContext = {}
     const hooks = createPlugin(instrumentationApi)
     const operationHooks = hooks.requestDidStart({})
     operationHooks.willSendResponse(responseContext)
-    t.equal(
-      operationSegment.name,
-      'GraphQL/operation/ApolloServer/<unknown>',
-      'should set operation to unknown'
-    )
+    t.equal(operationSegment.name, undefined, 'should set operation to unknown')
```
Comment (michaelgoin) [inline tests/unit/create-plugin.test.js:71]: This seems wrong. We should never have a segment with no name at all.

Comment (bizob2828) [reply tests/unit/create-plugin.test.js:71]: it's a unit test where `createSegment` is stubbed and where it would've named it accordingly

Comment (michaelgoin) [reply tests/unit/create-plugin.test.js:71]: oh i see, this is specifically testing this hook and this hook doesn't touch the segment name anymore.

sorry, i was getting this and the previous test change merged in my mind.

Comment (michaelgoin) [reply tests/unit/create-plugin.test.js:71]: i might recommend a `validationDidStart` equivalent to what was previously here (verifying named to unknown) but our versioned testing is so in-depth i know that is covered. so not critical.

**Review by michaelgoin — APPROVED** [review]
👍 Nice we can do this earlier naming now that we are using the document.

File: tests/unit/create-plugin.test.js (hunk 1 above)
Comment (bizob2828) [reply tests/unit/create-plugin.test.js:71]: yea I can do that. but it won't update it to unknown, it'll also be `undefined`.

Comment (bizob2828) [reply tests/unit/create-plugin.test.js:71]: actually. I'm going to change this test and add a different one that is asserting the default behavior of the segment name

**Review by michaelgoin — APPROVED** [review]
👍
